spec/integ/matrix-client-event-timeline.spec.js

```diff
@@ -0,0 +1,558 @@
+
+            var tl;
+            client.getEventTimeline(room, EVENTS[0].event_id
+            ).then(function(tl0) {
+                tl = tl0;
+                return client.paginateEventTimeline(tl, {backwards: true});
+            }).then(function(success) {
+                expect(success).toBeTruthy();
+                expect(tl.getEvents().length).toEqual(3);
+                expect(tl.getEvents()[0].event).toEqual(EVENTS[2]);
+                expect(tl.getEvents()[1].event).toEqual(EVENTS[1]);
+                expect(tl.getEvents()[2].event).toEqual(EVENTS[0]);
+                expect(tl.getPaginationToken(true)).toEqual("start_token1");
+                expect(tl.getPaginationToken(false)).toEqual("end_token0");
+            }).catch(exceptFail).done(done);
+
+            httpBackend.flush().catch(exceptFail);
+        });
+
+
+        it("should allow you to paginate forwards", function(done) {
+            var room = client.getRoom(roomId);
+
+            httpBackend.when("GET", "/rooms/!foo%3Abar/context/" +
+                             encodeURIComponent(EVENTS[0].event_id))
+                .respond(200, function() {
+                    return {
+                        start: "start_token0",
+                        events_before: [],
+                        event: EVENTS[0],
+                        events_after: [],
+                        end: "end_token0",
+                        state: [],
+                    };
+                });
+
+            httpBackend.when("GET", "/rooms/!foo%3Abar/messages")
+                .check(function(req) {
+                    var params = req.queryParams;
+                    expect(params.dir).toEqual("f");
+                    expect(params.from).toEqual("end_token0");
+                    expect(params.limit).toEqual(20);
+                }).respond(200, function() {
+                    return {
+                        chunk: [EVENTS[1], EVENTS[2]],
+                        end: "end_token1",
+                    };
+                });
+
+            var tl;
+            client.getEventTimeline(room, EVENTS[0].event_id
+            ).then(function(tl0) {
+                tl = tl0;
+                return client.paginateEventTimeline(tl, {backwards: false, limit: 20});
+            }).then(function(success) {
+                expect(success).toBeTruthy();
+                expect(tl.getEvents().length).toEqual(3);
+                expect(tl.getEvents()[0].event).toEqual(EVENTS[0]);
+                expect(tl.getEvents()[1].event).toEqual(EVENTS[1]);
+                expect(tl.getEvents()[2].event).toEqual(EVENTS[2]);
+                expect(tl.getPaginationToken(true)).toEqual("start_token0");
+                expect(tl.getPaginationToken(false)).toEqual("end_token1");
+            }).catch(exceptFail).done(done);
+
+            httpBackend.flush().catch(exceptFail);
+        });
+
+    });
+});
+
+// make the test fail, with the given exception
+function exceptFail(error) {
```

Comment:
well, not really. `except` as in 'exception'. Suggestions for better names appreciated. This little idiom turned out to be really useful in figuring out why failures were happening, instead of just getting "it didn't finish in 5 seconds".